spec/refresh_pools_spec.rb

```diff
@@ -1281,7 +1281,7 @@ def test_entitlement_regeneration
 
     # Yield to encapsulating test, applying any change it may have made
     sub = yield(owner, sub)
-    update_upstream_subscription(sub.id, sub)
+    sub = update_upstream_subscription(sub.id, sub)
```

Comment:
If we're always going to fetch the sub from upstream instead of trusting the block to manipulate it properly, it'd be better to just leave the job up to the container and do it all here.

Reply:
Removed. This was a forgotten bit.

Reply:
This was in reference to each of the tests below (1310, 1323, and 1336) all fetching the subscription from upstream every time instead of just returning the subscription as-is. The fetching into this update will end up being a slow no-op since any subscription changes between the two will be lost.

The existing code could be optimized a bit to check if there's even a difference to make to justify an update/refresh call, but at the very least we shouldn't be doing a fetch->update.

Reply:
Unfortunately, those fetches are needed for the same reason I added those assignments of product after adding content to it.

Without them the creation of subscription updates its products. This wasn't a problem before, because hosted resource didn't use DTOs so all attributes defaulted to null. Even collections. So now when we create a product, it has no content. We add a content, but the dto we use in subscription still has no content, so the content gets deleted during subscription creation. This was not a problem before because content was null so it was ignored during subscription creation.

As an alternative to fetching we could set those collections to null, but that would require more upkeep.

Reply:
Given that we aren't changing the product at all (even functionally), that should no-op. DTOs or not, the way the tests function, sending the subscription back up even with the field changed on the content that we've already changed upstream should end up doing nothing. If this is not the case, there's a regression in the hosted test logic.

However, my larger point here is that what this code is doing as a result of the changes to the individual tests is the following:

```
sub = get_upstream_subscription(subscription.id) # value returned from the block
update_upstream_subscription(sub.id, sub) # effective no-op since we'd be sending the sub we just fetched back up
```

Now, given how the tests actually ended up, the update itself is fairly extraneous. The remainder of the base function doesn't even use it, and if it's critical to the test, the test itself should probably be explicitly doing an update. That said, we could remove the entire subscription returning from the block, have the tests which want to update the subscription do their own update, and then drop the update in the base function entirely. The only silver lining of keeping it is that it seems to have exposed an issue we may not have caught otherwise.

Reply:
To recap discussion from IRC:
The issue you were seeing was due to the change to the content in two of the three tests being made on the legacy paths to provided and derived provided products. Once the target content path was updated to use the 4.0+ pathing, the update of the subscription no longer clobbered the change made in the test.

With that sorted, we can safely remove that whole semi-transparent update logic, since it's only used in two of the ten tests using the `test_entitlement_regeneration` function, and apparently caused a similar concern in two others. The final diff ends up looking as follows:

```diff
diff --git a/spec/refresh_pools_spec.rb b/spec/refresh_pools_spec.rb
index ac5c8f5bd..f0ae95d5d 100644
--- a/spec/refresh_pools_spec.rb
+++ b/spec/refresh_pools_spec.rb
@@ -1279,9 +1279,8 @@ describe 'Refresh Pools' do
 
     expect(concat_serials(entitlement, bonus_entitlement)).to eq(serial_concat)
 
-    # Yield to encapsulating test, applying any change it may have made
-    sub = yield(owner, sub)
-    update_upstream_subscription(sub.id, sub)
+    # Yield to encapsulating test
+    yield(owner, sub)
 
     # verify serial does not change on content update request that does not regenerate cert
     entitlement = @cp.get_entitlement(entitlement['id'])
@@ -1306,8 +1305,6 @@ describe 'Refresh Pools' do
       content = subscription['product']['productContent'].first['content']
       content.modifiedProductIds = [product_id2]
       update_upstream_content(content.id, content)
-
-      subscription
     }
   end
 
@@ -1316,11 +1313,9 @@ describe 'Refresh Pools' do
       product_id2 = random_string(nil, true)
       product2 = create_upstream_product(product_id2, { :name => 'test_prod_2' })
 
-      content = subscription['providedProducts'][0]['productContent'][0]['content']
+      content = subscription['product']['providedProducts'].first['productContent'].first['content']
       content.modifiedProductIds = [product_id2]
       update_upstream_content(content.id, content)
-
-      get_upstream_subscription(subscription.id)
     }
   end
 
@@ -1329,11 +1324,9 @@ describe 'Refresh Pools' do
       product_id2 = random_string(nil, true)
       product2 = create_upstream_product(product_id2, { :name => 'test_prod_2' })
 
-      content = subscription['derivedProvidedProducts'][0]['productContent'][0]['content']
+      content = subscription['product']['derivedProduct']['providedProducts'].first['productContent'].first['content']
       content.modifiedProductIds = [product_id2]
       update_upstream_content(content.id, content)
-
-      get_upstream_subscription(subscription.id)
     }
   end
 
@@ -1367,8 +1360,7 @@ describe 'Refresh Pools' do
       add_content_to_product_upstream(prov_product.id, content_id)
 
       subscription['product']['providedProducts'].push(prov_product)
-
-      subscription
+      update_upstream_subscription(subscription.id, subscription)
     }
 
     prov_product = json_body['products'].find {|p| p['id'] == prov_product_id}
@@ -1378,7 +1370,7 @@ describe 'Refresh Pools' do
   it 'regenerates entitlements when provided product is removed' do
     json_body, main_product = test_entitlement_regeneration { |owner, subscription|
       subscription['product']['providedProducts'] = []
-      subscription
+      update_upstream_subscription(subscription.id, subscription)
     }
 
     json_body['products'].size.should == 1
@@ -1389,8 +1381,6 @@ describe 'Refresh Pools' do
       content = subscription['product']['productContent'].first['content']
       content.label = 'shakeItOff'
       update_upstream_content(content.id, content)
-
-      subscription
     }
 
     product_json = json_body['products'].find {|p| p['id'] == main_product['id']}
@@ -1402,8 +1392,6 @@ describe 'Refresh Pools' do
       content = subscription['product']['productContent'].first['content']
       content.releaseVer = 'badBlood'
       update_upstream_content(content.id, content)
-
-      subscription
     }
 
     # releasever is not in json
@@ -1414,8 +1402,6 @@ describe 'Refresh Pools' do
       content = subscription['product']['productContent'].first['content']
       content.vendor = 'blankSpace'
       update_upstream_content(content.id, content)
-
-      subscription
     }
 
     product_json = json_body['products'].find {|p| p['id'] == main_product['id']}
@@ -1435,10 +1421,6 @@ describe 'Refresh Pools' do
       })
 
       add_content_to_product_upstream(product.id, content.id)
-
-      # Need to return the updated, upstream subscription here so we don't risk clobbering the addition.
-      # We shouldn't, anyway, but there's no need to risk it unnecessarily.
-      get_upstream_subscription(subscription.id)
     }
 
     product_json = json_body['products'].find {|p| p['id'] == main_product['id']}
@@ -1454,10 +1436,6 @@ describe 'Refresh Pools' do
       content = product['productContent'].first['content']
 
       remove_content_from_product_upstream(product.id, content.id)
-
-      # Need to return the updated, upstream subscription here so we don't risk clobbering the removal.
-      # We shouldn't, anyway, but there's no need to risk it unnecessarily.
-      get_upstream_subscription(subscription.id)
     }
 
     product_json = json_body['products'].find {|p| p['id'] == main_product['id']}
```